### ISSUE TITLE
When executing client.close(), onerror will be executed twice (Streamable HTTP)

### ISSUE THREAD

**Issue by shellRaining** [modelcontextprotocol/typescript-sdk#868]
**To Reproduce**
```typescript
import { Client } from "@modelcontextprotocol/sdk/client/index.js";
import { StreamableHTTPClientTransport } from "@modelcontextprotocol/sdk/client/streamableHttp.js";
import { log } from "console";

const client = new Client(
  {
    name: "comate",
    version: "1.0.0",
  },
  {
    capabilities: {},
  },
);
const transport = new StreamableHTTPClientTransport(
  new URL("https://api.githubcopilot.com/mcp/"),
  {
    requestInit: {
      headers: {
        Authorization: "Bearer xxx",
      },
    },
  },
);
client.onclose = () => {
  log(`close`);
};
client.onerror = (e) => {
  log(`error: ${e.message}`);
};
await client.connect(transport);
await new Promise((resolve) => {
  setTimeout(resolve, 1000);
});

await client.close()
```

run this code, will occur these log below

```
close
error: The operation was aborted.
error: The operation was aborted.
```

**Expected behavior**
only call onerror for once

**Comment by Anil567849** [comment]
I tried debugging the code and found that onerror is triggered only once. If my observation is incorrect, could you please provide more details on how to reproduce this issue?

**Comment by shellRaining** [comment]
https://github.com/user-attachments/assets/49b174d7-7043-409f-a37f-4b5fde6b1f91

my sdk version is 1.17.2, and notice that the github token should replace with yours, or will occur 401 error

**Comment by felixweinberger** [comment]
Port #880 to v2.

**Comment by windro-xdd** [comment]
I'd like to take this. I'll port the close-path fix to v2 so abort-triggered transport shutdown doesn't emit duplicate  callbacks, and add regression tests covering  behavior.

**Comment by azender1** [comment]
This feels like a broader execution-boundary issue more than just an event-handler bug.

Once the same logical close / teardown path can emit multiple times, downstream consumers have to guess whether they’re observing:

duplicate signaling
partial completion
or a genuinely new execution path

The pattern I’ve had the best results with is:

define a stable logical operation ID for the action
persist completion / terminal state at the boundary
make retries / repeats return that prior terminal state instead of re-emitting effects

Otherwise a duplicated callback can easily fan out into duplicated cleanup, logging, notifications, or external mutation downstream.

**Comment by pablopupo** [comment]
@windro-xdd are you still planning to work on this? There's been no linked PR since March, so if it's free I'll pick it up. The root cause matches the bot analysis above, the catch block fires onerror and the re-thrown error triggers the call-site catch a second time. I'd remove the duplicate call and add a regression test around client.close().

**Comment by windro-xdd** [comment]
Hey @pablopupo  please feel free to pick it up and go ahead with your proposed fix! I won't be available to work on this right now. Thanks for stepping in!
